proxy/src/http/websocket.rs

```diff
@@ -0,0 +1,255 @@
+        }
+    }
+}
+
+async fn serve_websocket(
+    websocket: HyperWebsocket,
+    config: &ProxyConfig,
+    cancel_map: &CancelMap,
+    session_id: uuid::Uuid,
+) -> anyhow::Result<()> {
+    let websocket = websocket.await?;
+    handle_client(config, cancel_map, session_id, WebSocketRW::new(websocket)).await?;
+    Ok(())
+}
+
+async fn ws_handler(
+    mut request: Request<Body>,
+    config: &'static ProxyConfig,
+    cancel_map: Arc<CancelMap>,
+    session_id: uuid::Uuid,
+) -> Result<Response<Body>, ApiError> {
+    let host = request.headers().get("host").and_then(|h| h.to_str().ok());
+
+    // Create a new proxy config with the hostname from the request.
+    let config = ProxyConfig {
+        tls_config: config.tls_config.clone(),
+        auth_backend: config.auth_backend.clone(),
+        secure_override_hostname: host
+            .and_then(|h| {
+                // Remove the port from the hostname, if present.
+                h.split(':').next()
+            })
+            .map(|h| h.to_string()),
+    };
+
+    // Check if the request is a websocket upgrade request.
+    if hyper_tungstenite::is_upgrade_request(&request) {
+        let (response, websocket) = hyper_tungstenite::upgrade(&mut request, None)
+            .map_err(|e| ApiError::BadRequest(e.into()))?;
+
+        tokio::spawn(async move {
+            if let Err(e) = serve_websocket(websocket, &config, &cancel_map, session_id).await {
+                error!("error in websocket connection: {:?}", e);
+            }
+        });
+
+        // Return the response so the spawned future can continue.
+        Ok(response)
+    } else {
+        json_response(StatusCode::OK, "Connect with a websocket client")
+    }
+}
+
+pub async fn task_main(
+    ws_listener: TcpListener,
+    config: &'static ProxyConfig,
+) -> anyhow::Result<()> {
+    scopeguard::defer! {
+        info!("websocket server has shut down");
+    }
+
+    let tls_config = config.tls_config.as_ref().map(|cfg| cfg.to_server_config());
+    let tls_acceptor: tokio_rustls::TlsAcceptor = match tls_config {
+        Some(config) => config.into(),
+        None => {
+            warn!("TLS config is missing, WebSocket Secure server will not be started");
+            return Ok(());
+        }
+    };
+
+    let tokio_listener = tokio::net::TcpListener::from_std(ws_listener)?;
+    let addr_incoming = AddrIncoming::from_listener(tokio_listener)?;
+
+    let tls_listener = TlsListener::new(tls_acceptor, addr_incoming).filter(|conn| {
+        if let Err(err) = conn {
+            error!("failed to accept TLS connection for websockets: {:?}", err);
+            ready(false)
+        } else {
+            ready(true)
+        }
+    });
+
+    let make_svc = hyper::service::make_service_fn(|_stream| async move {
+        Ok::<_, Infallible>(hyper::service::service_fn(
+            move |req: Request<Body>| async move {
+                let cancel_map = Arc::new(CancelMap::default());
```

Comment:
I wanted to make a global `cancel_map`, but Rust beat me. Also I'm not sure what `cancel_map` does actually, so would be nice to have someone familiar with proxy to comment here.

Reply:
When the connection is established postgres sends special cancel_key to client. Using this key client can open a new connection and call `select pg_cancel_backend($cancel_key)` to kill currently running query in the main connection. Proxy keeps a map to be able to route that cancel connection.

Usually clients call cancel thing on various timeouts or signals. Not sure is js driver supports it at all.